File: src/jsonrpc/eth/newBlockFilter/eth_newBlockFilter.zig
const std = @import("std");
const types = @import("../../types.zig");

/// Creates a filter in the node, allowing for later polling. Registers client interest in new blocks, and returns an identifier.
///
/// Example:
/// Result: "0x01"
///
/// Implements the `eth_newBlockFilter` JSON-RPC method.
pub const EthNewBlockFilter = @This();

/// The JSON-RPC method name
pub const method = "eth_newBlockFilter";

/// Parameters for `eth_newBlockFilter`
pub const Params = struct {
    pub fn jsonStringify(self: Params, jws: *std.json.Stringify) !void {
        _ = self;
        try jws.write(.{});
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Params {
        _ = allocator;
        _ = source;
        _ = options;
        return Params{};
    }
};

/// Result for `eth_newBlockFilter`
pub const Result = struct {
    /// hex encoded unsigned integer
    value: types.Quantity,

    pub fn jsonStringify(self: Result, jws: *std.json.Stringify) !void {
        try jws.write(self.value);
    }

    pub fn jsonParseFromValue(allocator: std.mem.Allocator, source: std.json.Value, options: std.json.ParseOptions) !Result {
        return Result{
            .value = try std.json.innerParseFromValue(types.Quantity, allocator, source, options),
        };
    }
};

File: src/jsonrpc/eth/newBlockFilter/eth_newBlockFilter.go
package eth

import (
	"encoding/json"
	"fmt"

	"github.com/ethereum/execution-apis/types"
)

// Creates a filter in the node, allowing for later polling. Registers client interest in new blocks, and returns an identifier.
//
// Example:
// Result: "0x01"
//
// Implements the eth_newBlockFilter JSON-RPC method.

// Method is the JSON-RPC method name
const Method = "eth_newBlockFilter"

// Params represents the parameters for eth_newBlockFilter
type Params struct {
}

// MarshalJSON implements json.Marshaler for Params.
// JSON-RPC 2.0 uses positional array parameters.
func (p Params) MarshalJSON() ([]byte, error) {
	return json.Marshal([]interface{}{
	})
}

// UnmarshalJSON implements json.Unmarshaler for Params.
func (p *Params) UnmarshalJSON(data []byte) error {
	var arr []json.RawMessage
	if err := json.Unmarshal(data, &arr); err != nil {
		return err
	}

	if len(arr) != 0 {
		return fmt.Errorf("expected 0 parameters, got %d", len(arr))
	}

	return nil
}

// Result represents the result for eth_newBlockFilter
//
// hex encoded unsigned integer
type Result struct {
	Value types.Quantity `json:"-"`
}

// MarshalJSON implements json.Marshaler for Result.
func (r Result) MarshalJSON() ([]byte, error) {
	return json.Marshal(r.Value)
}

// UnmarshalJSON implements json.Unmarshaler for Result.
func (r *Result) UnmarshalJSON(data []byte) error {
	return json.Unmarshal(data, &r.Value)
}
